python/cudf/cudf/core/dataframe.py

```diff
@@ -4265,8 +4286,64 @@ def from_records(self, data, index=None, columns=None, nan_as_null=False):
                 df[k] = Series(data[k], nan_as_null=nan_as_null)
 
         if index is not None:
-            indices = data[index]
-            return df.set_index(indices.astype(np.int64))
+            if is_scalar(index):
+                df = df.set_index(index)
+            else:
+                df._index = as_index(index)
+        return df
+
+    @classmethod
+    def _from_arrays(self, data, index=None, columns=None, nan_as_null=False):
+        """Convert a numpy/cupy array to DataFrame.
+
+        Parameters
+        ----------
+        data : numpy/cupy array of ndim 1 or 2,
+            dimensions greater than 2 are not supported yet.
+        index : Index or array-like
+            Index to use for resulting frame. Will default to
+            RangeIndex if no indexing information part of input data and
+            no index provided.
+        columns : list of str
+            List of column names to include.
+
+        Returns
+        -------
+        DataFrame
+        """
+
+        data = cupy.asarray(data)
+        if data.ndim != 1 and data.ndim != 2:
+            raise ValueError(
+                "records dimension expected 1 or 2 but found {!r}".format(
+                    data.ndim
+                )
+            )
+
+        if data.ndim == 2:
+            num_cols = len(data[0])
+        else:
+            # Since we validate ndim to be either 1 or 2 above,
+            # this case can be assumed to be ndim == 1.
+            num_cols = 1
+
+        if columns is None:
+            names = [i for i in range(num_cols)]
+        else:
+            if len(columns) != num_cols:
+                msg = "columns length expected {!r} but found {!r}"
+                raise ValueError(msg.format(num_cols, len(columns)))
+            names = columns
+
+        df = DataFrame()
+        if data.ndim == 2:
+            for i, k in enumerate(names):
+                df[k] = Series(data[:, i], nan_as_null=nan_as_null)
+        elif data.ndim == 1:
+            df[names[0]] = Series(data, nan_as_null=nan_as_null)
```

Comment:
Can we construct using columns to avoid unnecessary index alignment checking / handling?